src/Build/Xml/XmlReaderExtension.cs

```diff
@@ -34,12 +37,22 @@ private XmlReaderExtension(string file)
                 _stream = new FileStream(file, FileMode.Open, FileAccess.Read, FileShare.Read);
                 _streamReader = new StreamReader(_stream, Encoding.UTF8, true);
                 Reader = GetXmlReader(_streamReader, out _encoding);
+                var bom = EncodingUtilities.StreamStartsWithBom(_stream);
 
                 // Override detected encoding if xml encoding attribute is specified
                 var encodingAttribute = Reader.GetAttribute("encoding");
                 _encoding = !string.IsNullOrEmpty(encodingAttribute)
                     ? Encoding.GetEncoding(encodingAttribute)
                     : _encoding;
+
+                // If the encoding is UTF8 but there was not a BOM detected in the file then we need
+                // to override the detected Encoding with UTF8 and encoderShouldEmitUTF8Identifier
+                // set to false. This is to prevent the BOM from being added if we save the file
+                // later using the detected encoding.
+                if (Equals(_encoding, Encoding.UTF8) && !bom)
```

Comment:
Help me understand why you can't sniff the encoding off the streamreader which does the right thing here?